Cognite.Simulator.Utils/ModelLibraryBase.cs

```diff
@@ -234,23 +224,96 @@ private async Task<T> TryReadRemoteModelRevision(string modelRevisionExternalId,
-                    && s.IsExtracted
-                    && !string.IsNullOrEmpty(s.FilePath))
+                .Where(state => state.Id == modelRevision.Id.ToString())
                 .OrderByDescending(s => s.CreatedTime);
 
-            var model = modelVersions.FirstOrDefault();
+            var modelState = modelVersions.FirstOrDefault();
 
-            if (model != null)
+            if (modelState == null)
             {
-                return model;
+                modelState = UpsertModelRevisionInState(modelRevision);
             }
 
-            return await TryReadRemoteModelRevision(modelRevisionExternalId, CancellationToken.None).ConfigureAwait(false);
+            if (modelState == null)
+            {
+                _logger.LogError("Failed to get model state for {ModelRevisionExternalId}", modelRevisionExternalId);
+                return null;
+            }
+
+            UpdateModelParsingInfo(modelState, modelRevision);
+
+            var downloaded = modelState.Downloaded;
+
+            if (!downloaded && modelState.DownloadAttempts < _config.MaxDownloadAttempts)
+            {
+                downloaded = await DownloadFileAsync(modelState).ConfigureAwait(false);
+            }
+
+            if (downloaded && modelState.ShouldProcess())
+            {
+                await ExtractModelInformationAndPersist(modelState, token).ConfigureAwait(false);
+            }
+
+            return modelState;
+        }
+
+        /// <inheritdoc/>
+        public Task<T> GetModelRevision(string modelRevisionExternalId)
+        {
+            return GetOrAddModelRevision(modelRevisionExternalId);
+        }
+
+        /// <summary>
+        /// This method gives a safe way to get a processed model revision.
+        /// Internally, it uses a Lazy Task to ensure that only one thread processes a given model revision at a time.
+        /// </summary>
+        /// <param name="modelRevisionExternalId">Model revision External ID</param>
+        /// <param name="remoteRevision">Remote model revision object, if available, otherwise it will be fetched from CDF</param>
+        private async Task<T> GetOrAddModelRevision(string modelRevisionExternalId, SimulatorModelRevision remoteRevision = null, CancellationToken token = default)
+        {
+            var lazyTask = _revisionsTasks.GetOrAdd(modelRevisionExternalId, id =>
+                new Lazy<Task<T>>(async () =>
+                {
+                    // It will only ever be executed ONCE for a given key, by the first thread that accesses ".Value".
+                    try
+                    {
+                        return await GetOrAddModelRevisionImpl(id, remoteRevision, token).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        _logger.LogDebug("Processing finished. Removing processing task for model revision {ExternalId}", id);
+                        _revisionsTasks.TryRemove(id, out _);
```

Comment:
I've been looking at this since the last few minutes and I think there might be an issue here. Not 100% sure though.

There's this tiny window just at the end of task completion here that can be a bit problematic. 

1: T1: Starts with externalId 1
2: T2: Starts with externalId 1
3: T1: Does its thing and reaches the remove line
4: T1: Removes the task and finishes
5: T2: Finds that there's no task with id 1, it creates a new task for the same id and starts executing it again leading to a duplicate operation.

I might be wrong, so ignore this comment if this does not make any sense.

Reply:
there is a test for this. it's exaclt what this PR solves

Reply:
Made this more sequencial now, so the worst case is that next call (if unfortunate) will call `GetOrAddModelRevisionImpl`, which will exit after `TryReadRemoteModelRevision` if no work needed (nothing got updated)